Use Conan for installing dependencies

SJaffa opened the pull request:
Starting from the development branch, first issue was the same spdlog error as in [sopt](https://github.com/astro-informatics/sopt/pull/265/files)

- [x] Update install instructions in README.md with https://github.com/astro-informatics/purify/pull/302#issuecomment-1532985806
- [x] Fix code errors in algo_factory.h
- [x] Only include sopt TF headers from algo_factory.h when `cppflow==on` option is passed to cmake
- [x] In sopt, move `class GProximal` to the `sopt::algorithm` namespace
- [ ] Update documentation GitHub Actions workflow

SJaffa committed:
Fixed spdlog error, same as sopt

SJaffa committed:
Fix error in CI

SJaffa committed:
Add conan install steps to workflows

SJaffa committed:
Conanfile with same dependencies as sopt

SJaffa committed:
Added some of the lookup_package packages into Conan

SJaffa committed:
Fixed linking and versions

tkoskela committed:
Update build to use conan. Update to conanfile.py similar to sopt.

SJaffa commented:
- Checks fail
	- All Ubuntu fails in build step because of spdlog error in SOPT, need to pull SOPT to fix this
	- macOS clang also fails in build step because of spdlog error
	- macOS gcc fails in 'install conan on gcc' 
	    - boost fiber error:

`boost/1.78.0: WARN: Boost component 'fiber' is missing libraries. Try building boost with '-o boost:without_fiber'. (Option is not guaranteed to exist)
ERROR: boost/1.78.0: Error in package_info() method, line 1592
	raise ConanException("These libraries were expected to be built, but were not built: {}".format(non_built))
	ConanException: These libraries were expected to be built, but were not built: {'boost_fiber'}`

tkoskela committed:
Fix class name

tkoskela committed:
Apply changes done in sopt

tkoskela commented:
We need to close https://github.com/astro-informatics/sopt/issues/279 before we can make progress with this

tkoskela committed:
WIP: Builds object files but fails during linking

tkoskela committed:
A version that compiles and passes tests. Outstanding thigs:

- How do we build the sopt package automatically? Right now it points
to my private test version of the package.
- tests/measurement_operator.cc had some lines with semicolons missing!
- in tests/mpi_algo_factory.cc I had to comment out statements with
CAPTURE(uv_data.vis.head(5)); because something was going wrong with
the macro expansion.
- the mpi_algo_factory test only passes when the number of processes
is 2. Otherwise it seems to hang indefinitely. There is a lot of code
around world.size == 2.

tkoskela committed:
Checkout sopt and create package to install purify

tkoskela committed:
Skip broken symlinks, remove unnecessary options

tkoskela committed:
Fix typo

tkoskela committed:
Add path to sopt repo so it doesn't override purify

tkoskela committed:
Set config with conan command

tkoskela committed:
Define env variables inside the job

tkoskela committed:
Add debugging session dispatch

tkoskela committed:
Move tmate up

tkoskela committed:
Set True as string

tkoskela committed:
pass build settings and options to create

tkoskela committed:
Re-add build missing flag

tkoskela committed:
Add sopt submodule

tkoskela committed:
Merge branch 'tk/submodule' into tk/port-from-sopt

tkoskela committed:
checkout with sopt submodule

SJaffa committed:
Remove regresssions option from conanfile

SJaffa committed:
Update buld instructions to use conan

SJaffa commented:
@tkoskela is this the branch you got the SOPT package working on? This one fails for me as it can't find Eigen and spdlog.

tkoskela committed:
Merge branch 'sj/conan-build' into tk/port-from-sopt

tkoskela committed:
Merge pull request #304 from astro-informatics/tk/port-from-sopt

Port build system improvements from sopt

tkoskela committed:
Merge updates from sopt workflow

tkoskela committed:
Update sopt to latest

tkoskela commented:
The algo_factory that creates the `ImagingForwardBackward` object in purify is templated by the algorithm type https://github.com/astro-informatics/purify/blob/5833dbd35f214212aa17c368984bcae180cf5dee/cpp/main.cc#L440 but the template parameter is checked by an if statement and there's a different branch for each algorithm class https://github.com/astro-informatics/purify/blob/5833dbd35f214212aa17c368984bcae180cf5dee/cpp/purify/algorithm_factory.h#L142

:zany_face:

tkoskela committed:
Always override zlib version to avoid conflicts

tkoskela committed:
Add support for learned g_proximal in forward-backward algorithm

SJaffa commented:
The Cmake flags for building SOPT are specified in [cmakefiles/LookUpSopt.cmake](https://github.com/astro-informatics/purify/blob/0a940e409e95bfbddf4a9d74daa9daad56938165/cmake_files/LookUpSopt.cmake#L34) so we can add an option in the `conanfile.py` to turn `cppflow` on and off and pass that through to SOPT like they do with `openmp` and `logging`

SJaffa committed:
Adding cppflow flag to purify and passing through to sopt

SJaffa committed:
Linting

SJaffa committed:
More linting

SJaffa committed:
More linting

tkoskela committed:
Delete lookup files for dependencies handled by conan

tkoskela committed:
Update to c++17

tkoskela committed:
Work in progress on building with conan packages

tkoskela commented:
TODO: Add these instructions to README

To build `purify` using this PR and https://github.com/astro-informatics/sopt/pull/350

- [Install tensorflow C api](https://serizba.github.io/cppflow/installation.html)
- git clone the UCL fork of cppflow and create a conan package using
```
git clone git@github.com:UCL/cppflow.git
conan create ./cppflow/ -pr:h=default -pr:b=default
```
- git clone sopt and create a conan package using 
```
git clone git@github.com:astro-informatics/sopt.git
conan create ./sopt/ --build missing -s compiler.libcxx=libstdc++11 -o cppflow=on  -pr:h=default -pr:b=default
```
- conan install and build as usual
```
conan install .. --build missing -o cppflow=on -pr:h=default -pr:b=default
conan build ..
```

Commit:
Fix g_proximal_type naming

Commit:
Fix bracket error and g_proximal type naming

SJaffa committed:
Updating installation instructions to use conan packages

Commit:
Access Psi through g_proximal for all algorithms

SJaffa committed:
Reordered tests to be alphabetical

SJaffa committed:
Reordered mpi tests

SJaffa committed:
Merge branch 'mm_use_gproximal_base' into sj/reorder-tests

Commit:
Fix g_proximal constructors to be exclusive

tkoskela committed:
Speed up sopt package creation

tkoskela committed:
Merge branch 'sj/conan-build' of github.com:astro-informatics/purify into sj/conan-build

Commit:
Set MPI communicators for FB algorithm

Commit:
Fix non-MPI tests when compiled with MPI

mmcleod89 committed:
Merge pull request #315 from astro-informatics/mm_use_gproximal_base

Make purify code compatible with refactored sopt

tkoskela committed:
remove cmake shell for testing

tkoskela committed:
Update gcc version

tkoskela committed:
clean up commented code

tkoskela committed:
Merge pull request #319 from tkoskela/sj/conan-build

Fix parsing error and update gcc version

tkoskela committed:
Add guard around including sopt cppflow header

tkoskela committed:
Merge branch 'sj/conan-build' of github.com:astro-informatics/purify into sj/conan-build

SJaffa committed:
Update sopt submodule to latest dev branch

Commit:
Set Primal Dual tests to "should fail"

krishnakumarg1984 committed:
feat: adds gh action to facilitate use of latest clang-format

tkoskela committed:
Modify inplace to get meaningful git diff

tkoskela committed:
Create .clang-format-ignore

tkoskela committed:
Merge pull request #322 from astro-informatics/kg/update_ghactions_clangformat_version

adds gh action to facilitate use of latest clang-format

tkoskela committed:
Change back to correct format

krishnakumarg1984 committed:
chore: reformat project files according to clang-format 16

tkoskela committed:
Require standard version and disable extensions

tkoskela committed:
Merge branch 'sj/conan-build' of github.com:astro-informatics/purify into sj/conan-build

tkoskela committed:
Install boost without fiber, only run gcc mac workflow

tkoskela committed:
Clean up packages installed by conan from brew and apt install

tkoskela committed:
Add comments and exclude python

tkoskela committed:
Re-populate the matrix

tkoskela committed:
Drop support for OpenMP on apple-clang

tkoskela committed:
Follow cppreference recommendation for initialising arrays

tkoskela committed:
Include array suggested on stackoverflow

https://stackoverflow.com/a/42846010

tkoskela committed:
Initialise with single braces

tkoskela committed:
Merge pull request #327 from astro-informatics/tk/fix-apple-clang

Fix/work-around errors in apple-clang

tkoskela committed:
Merge pull request #326 from astro-informatics/tk/debug-boost-build

Install boost without fiber and python

tkoskela commented on .github/workflows/cmake.yml:
```diff
@@ -29,51 +41,96 @@ jobs:
+      id: conan
+      uses: turtlebrowser/get-conan@main
+      with:
+        version: 1.59.0
```
```suggestion
        version: 1.60.1
```

tkoskela commented on .github/workflows/documentation.yml:
```diff
@@ -3,15 +3,17 @@ name: Documentation
-        run: sudo apt install libfftw3-dev libtiff5-dev libboost-all-dev libeigen3-dev libyaml-cpp-dev libcfitsio-dev casacore-dev doxygen graphviz
-        
+        run: |
+            sudo apt install libfftw3-dev libtiff5-dev libboost-all-dev libeigen3-dev libyaml-cpp-dev libcfitsio-dev casacore-dev doxygen graphviz
+            pip install conan
```
We should clean up the apt installed packages, similar to cmake.yml. Also install conan with the action

tkoskela replied:
It looks like the documentation workflow needs more updating. It's trying to build with `cmake` which will fail I assume.

SJaffa replied:
I've updated it to match the other workflows and copied the doxygen workaround we used in `sopt` but I can't test if it works as it only runs when we push to `development`.

tkoskela commented on .github/workflows/linting.yml:
```diff
@@ -3,28 +3,27 @@ name: Linting
     runs-on: ubuntu-20.04
     steps:
       - name: Checkout repo
         uses: actions/checkout@v2
```
```suggestion
        uses: actions/checkout@v3
```

tkoskela commented on cmake_files/LookUpSopt.cmake:
Sopt gets installed by conan, so we should delete this file?

tkoskela commented on cmake_files/dependencies.cmake:
```diff
@@ -1,19 +1,42 @@
+  message(FATAL_ERROR "Eigen is required")
+else()
+  message("eigen_FOUND: ${eigen_FOUND}")
+  message("Eigen3_FOUND: ${Eigen3_FOUND}")
```
This was for debugging and can be deleted
```suggestion

```

tkoskela commented on cmake_files/dependencies.cmake:
I think all the `find_package` calls should have `REQUIRED` and we can drop the manual error handling

mmcleod89 replied:
I think we need to resolve this before it will let me merge; are we happy to move this to a different issue given that it is currently building?

20DM replied:
Happy to follow up on this in a separate merge request (will open an issue to remember).

tkoskela commented on conanfile.py:
```diff
@@ -0,0 +1,133 @@
+
+    def requirements(self):
+        # To prevent a conflict in the version of zlib required by libtiff and
+        # doxygen, override the version of zlib when either of them is required
```
```suggestion
        # Override version of zlib to prevent a conflict between versions of zlib required by depedencies
```

tkoskela commented on conanfile.py:
```diff
@@ -0,0 +1,133 @@
+    #   # TODO: Switch this on for Debug, off for Release
+    #   cmake.definitions['CMAKE_VERBOSE_MAKEFILE:BOOL'] = "ON"
+    #   cmake.configure()
+    #   cmake.build()
```
```suggestion

```

tkoskela commented on conanfile.txt:
conanfile.txt is not used in favour of conanfile.py, this should be deleted

tkoskela commented on cpp/tests/CMakeLists.txt:
```diff
@@ -1,19 +1,19 @@
   add_library(common_catch_main_object OBJECT "common_catch_main.cc")
-  if(SPDLOG_INCLUDE_DIR)
-    target_include_directories(common_catch_main_object SYSTEM PUBLIC ${SPDLOG_INCLUDE_DIR})
+  if(spdlog_INCLUDE_DIR)
```
This seems like the wrong way to check if we need spdlog, if that's what it's doing. Shouldn't this be `spdlog_FOUND` or `logging`?

Similar comments to `CATCH` and `sopt` below

tkoskela replied:
I don't really understand the logic of having these if statements, but I'm going to leave them alone for now. If the include dirs aren't found they don't get included in the target, but surely the build will then fail? Is it possible that they have been provided through some other way?

tkoskela commented on cpp/tests/algo_factory.cc:
```diff
@@ -73,7 +73,7 @@ TEST_CASE("padmm_factory") {
 }
 
-TEST_CASE("primal_dual_factory") {
+TEST_CASE("primal_dual_factory", "[!shouldfail]") {
```
This probably needs an explanation from #317

tkoskela commented on cpp/tests/algo_factory.cc:
```diff
@@ -109,7 +109,7 @@ TEST_CASE("primal_dual_factory") {
+          imsizey, imsizex, sara.size(), 20, true, true, 1e-2, 1, op_norm);
 
   auto const diagnostic = (*primaldual)();
   CHECK(diagnostic.niters == 16);
```
Doesn't really matter since we expect it to fail, but I would still remove the check `niters == 16`

tkoskela commented on cpp/tests/mpi_algo_factory.cc:
```diff
@@ -45,7 +45,7 @@ TEST_CASE("Serial vs. Serial with MPI PADMM") {
   auto uv_data = dirty_visibilities({input_data_path}, world);
   uv_data.units = utilities::vis_units::radians;
-  if (world.is_root()) CAPTURE(uv_data.vis.head(5));
+  // if (world.is_root()) CAPTURE(uv_data.vis.head(5));
```
I remember commenting these out because they were causing build errors, maybe worth checking if I can fix them :blush:

tkoskela committed:
Clarify Readme

- Organised dependencies to user-provided and automatically downloaded
- Updated dependency versions
- Split build instructions into vanilla and TF versions
- Added disclaimer on Docker

tkoskela committed:
Remove trusty references and some extra whitespace

tkoskela committed:
Make path to sopt consistent

tkoskela committed:
Apply suggestions from code review

Adding the straightforward changes

tkoskela committed:
Update cmake_files/dependencies.cmake

tkoskela committed:
removed old conanfile

tkoskela committed:
Merge branch 'sj/conan-build' of github.com:astro-informatics/purify into sj/conan-build

mmcleod89 committed:
Fix PURIFY_CPPFLOW flag

Following the pattern of other flags in this file; PURIFY_CPPFLOW doens't get set properly without this line, I think it is otherwise not defined.

Commit:
remove checks for n_iters from tests

mmcleod89 commented:
Looking at things here I think that this is up to date now with what I expect, keeping the forward-backward tensorflow test out of it for a separate PR. 👍

tkoskela committed:
Add comments to explain shouldfail test

tkoskela committed:
Catch2 macro does not like one-line if statements

tkoskela committed:
Add missing step

tkoskela commented:
Ok, from my point of view the only outstanding item is [updating the documentation workflow](https://github.com/astro-informatics/purify/pull/302#discussion_r1261404711). I think everything else looks good to go.

SJaffa committed:
Merge pull request #312 from astro-informatics/sj/reorder-tests

Make tests in alphabetical order

SJaffa commented on README.md:
````diff
@@ -17,104 +17,139 @@ This documentation outlines the necessary and optional [dependencies](#dependenc
 
-Alternatively, you can build **PURIFY** entirely from the source code. Once the mandatory dependencies are present, `git clone` from the [GitHub repository](https://github.com/astro-informatics/purify):
+    ``` bash
+    git clone --recurse-submodules https://github.com/astro-informatics/purify.git
````
This step created the `sopt` folder for me but didn't actually fetch the files, I had to add a `git submodule update --init --recursive` step. Is this the same for you @tkoskela @20DM ?

tkoskela replied:
I think this is because it does not recurse submodules in all branches, only the one it pulls when you clone. If I do
```
git clone --recurse-submodules --branch sj/conan-build https://github.com/astro-informatics/purify.git
```
I get all the submodules included, so I assume once this is merged into the default branch the current documentation will be fine.

SJaffa commented on README.md:
````diff
@@ -17,104 +17,139 @@ This documentation outlines the necessary and optional [dependencies](#dependenc
 
-Then, the program can be built with standard `CMake` command:
+    ```bash
+    conan create /path/to/purify/sopt/ --build missing -s compiler.libcxx=libstdc++11 -pr:h=default -pr:b=default
````
This gives an error about broken symlinks, which can be fixed by editing your `.conan/conan.conf` but did we find a way to pass that option on the command line?

tkoskela replied:
Ah, I must have had that in my conf. That's a bit annoying. You could set an [environment varible](https://docs.conan.io/en/1.46/reference/env_vars.html#conan-skip-broken-symlinks-check)

SJaffa replied:
I've added a note to the readme on how to fix this error.

SJaffa commented on README.md:
````diff
@@ -17,104 +17,139 @@ This documentation outlines the necessary and optional [dependencies](#dependenc
+
+    ``` bash
+    cd /path/to/purify
+    mkdir build
+    conan install .. --build missing -o cppflow=on -pr:h=default -pr:b=default
````
```suggestion
    mkdir build
    cd build
    conan install .. --build missing -o cppflow=on -pr:h=default -pr:b=default
```

SJaffa replied:
Either need to be in the build folder or change the `..` to `.` in both the `conan install` and `conan build` commands

SJaffa commented on README.md:
```diff
@@ -17,104 +17,139 @@ This documentation outlines the necessary and optional [dependencies](#dependenc
 saved and time-stamped. Additionally, a config file with the settings used will
 be saved and time-stamped, helping for reproducibility and book-keeping.
 
 Docker
```
Should we remove this section since we have not maintained the Docker image? This isn't really related to this PR, just a general update so I'll make a separate issue

tkoskela replied:
My assumption was that if there is a self-contained dockerfile on dockerhub it might still work, hence the disclaimer I added below. I agree, create a separate issue and remove it from the documentation if the container is unusable.

SJaffa replied:
I guess this isn't relevant to this PR so I'll make an issue and we can leave it for now, then I think this one is ready to merge!

SJaffa commented on README.md:
````diff
@@ -17,104 +17,139 @@ This documentation outlines the necessary and optional [dependencies](#dependenc
 ```
-cd /path/to/code/build
+cd /path/to/purify/build
 ctest .
````
All tests are passing when building with or without `cppflow` but we don't have a test in Purify that checks `cppflow`, only is `SOPT`

tkoskela replied:
This is addressed in #324

tkoskela committed:
Update README.md

Co-authored-by: S Jaffa <SJaffa@users.noreply.github.com>

SJaffa committed:
Updated docs workflow

SJaffa committed:
Passing docs option into Sopt

SJaffa committed:
Merge branch 'sj/conan-build' of https://github.com/astro-informatics/purify into sj/conan-build

SJaffa commented on README.md:
````diff
@@ -17,104 +17,139 @@ This documentation outlines the necessary and optional [dependencies](#dependenc
+    ```bash
+    conan create /path/to/purify/sopt/ --build missing -s compiler.libcxx=libstdc++11 -pr:h=default -pr:b=default
+    ```
 
````
```suggestion
If you get an error about broken symlinks you can set `skip_broken_symlinks_check = True` in your `~/.conan/conan.conf` file or [set an environment variable](https://docs.conan.io/en/1.46/reference/env_vars.html#conan-skip-broken-symlinks-check)
```

SJaffa committed:
Add note on symlinks

SJaffa commented:
I've checked the installation instructions in the readme work for me from a clean setup, I've updated the documentation workflow (although not tested this as it only runs when we push to development branch) and the other couple of outdated bits in the readme I've made new issues as they can be solved separately. So if someone could check over the docs workflow I think this is finally ready to merge! (assuming the tests pass...)

mmcleod89 reviewed (approved):
C++ code is good for me; as far as the docs action goes it looks okay to me but I don't think it's possible to run it manually so I'm happy for there to be a separate issue if there needs to be a further tweak there rather than extending this PR any further.
